core/cmd/shell_local.go

```diff
@@ -1069,41 +1037,143 @@ func (s *Shell) RemoveBlocks(c *cli.Context) error {
 	}
 
 	lggr := logger.Sugared(s.Logger.Named("RemoveBlocks"))
-	ldb := pg.NewLockedDB(cfg.AppID(), cfg.Database(), cfg.Database().Lock(), lggr)
 	ctx, cancel := context.WithCancel(context.Background())
 	go shutdown.HandleShutdown(func(sig string) {
 		cancel()
 		lggr.Info("received signal to stop - closing the database and releasing lock")
 
-		if cErr := ldb.Close(); cErr != nil {
-			lggr.Criticalf("Failed to close LockedDB: %v", cErr)
-		}
+		s.cleanupDBAndLogger(lggr)
+	})
+
+	app, err := s.AppFactory.NewApplication(ctx, s.Config, s.Logger, s.Registerer, s.DS, s.KeyStore)
+	if err != nil {
+		return s.errorOut(errors.Wrap(err, "fatal error instantiating application"))
+	}
+
+	err = app.DeleteLogPollerDataAfter(ctx, chainID, start)
+	if err != nil {
+		return s.errorOut(err)
+	}
+
+	lggr.Infof("RemoveBlocks: successfully removed blocks")
+
+	return nil
+}
+
+// BeforeNode initializes database, keystore, logger, and beholder for node startup.
+// This is used as a Before hook in CLI commands that require these components.
+func (s *Shell) BeforeNode(c *cli.Context) error {
+	if err := s.initializeComponents(c); err != nil {
+		return s.errorOut(err)
+	}
+	return nil
+}
 
-		if cErr := s.CloseLogger(); cErr != nil {
-			log.Printf("Failed to close Logger: %v", cErr)
+// cleanupDBAndLogger is a thread-safe helper method to close the database and logger.
+// This is used in multiple places: shutdown handler, signal handler, and cleanup.
+// It uses sync.Once to ensure cleanup happens exactly once, even if called concurrently.
+func (s *Shell) cleanupDBAndLogger(lggr logger.SugaredLogger) {
+	s.CleanupOnce.Do(func() {
+		if s.LDB != nil {
+			if err := s.LDB.Close(); err != nil {
+				lggr.Criticalf("Failed to close LockedDB: %v", err)
+			}
+			s.LDB = nil
+			lggr.Debug("Closed DB")
+		}
+		if s.CloseLogger != nil {
+			if err := s.CloseLogger(); err != nil {
+				log.Printf("Failed to close Logger: %v", err)
+			}
+			s.CloseLogger = nil
+			log.Printf("Closed logger")
 		}
```

Comment:
Is there actually a chance of these being nil?
```suggestion
		if err := s.LDB.Close(); err != nil {
			lggr.Criticalf("Failed to close LockedDB: %v", err)
		}
		if err := s.CloseLogger(); err != nil {
			log.Printf("Failed to close Logger: %v", err)
		}
```
I think we can skip the extra logs. Our goal here should be as little change as possible.

Reply:
Done